src/adaptation/precond.jl

```diff
@@ -2,389 +2,227 @@
+    is_update::Bool=true
+)
+    resize!(ve, θ)
+    push!(ve, θ)
+    if ve.n >= ve.n_min && is_update
+        ve.var .= getest(ve)
+    end
+end
+
+# NOTE: this naive variance estimator is used only in testing
+mutable struct NaiveVar{E<:AbstractVecOrMat{<:AbstractFloat},T<:AbstractVector{E}} <: VarEstimator{T}
     n :: Int
     S :: T
 end
 
-NaiveCov(::Type{T}, sz::Union{Tuple{Int}}) where {T<:AbstractFloat} = NaiveCov(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Vector{T}}())
+NaiveVar(::Type{T}, sz::Tuple{Int,Int}) where {T<:AbstractFloat} = NaiveVar(0, Vector{Matrix{T}}())
 
-function add_sample!(nc::NaiveCov, s::AbstractVector)
-    nc.n += 1
-    push!(nc.S, s)
-end
+NaiveVar(sz::Tuple{Vararg{Int}}; kwargs...) = NaiveVar(Float64, sz; kwargs...)
 
-function get_cov(nc::NaiveCov)
-    @assert nc.n >= 2 "Cannot get covariance with only one sample"
-    return Statistics.cov(nc.S)
+function Base.push!(nv::NaiveVar, s::AbstractVecOrMat)
+    nv.n += 1
+    push!(nv.S, s)
 end
 
-function reset!(nc::NaiveCov{T}) where {T}
-    nc.n = 0
-    nc.S = T()
+function reset!(nv::NaiveVar{T}) where {T}
+    nv.n = 0
+    nv.S = T()
 end
 
-# Ref: https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_covar_estimator.hpp
-mutable struct WelfordCov{T<:AbstractFloat} <: CovEstimator{T}
-    n :: Int
-    μ :: Vector{T}
-    M :: Matrix{T}
-    δ :: Vector{T} # temporary
+function getest(nv::NaiveVar)
+    @assert nv.n >= 2 "Cannot estimate variance with only one sample"
+    return Statistics.var(nv.S)
 end
 
-function WelfordCov(::Type{T}, sz::Tuple{Int}) where {T<:AbstractFloat}
-    d, = sz
-    WelfordCov(0, zeros(T, d), zeros(T, d, d), zeros(T, d))
-end
-
-function reset!(wc::WelfordCov{T}) where {T<:AbstractFloat}
-    wc.n = 0
-    wc.μ .= zero(T)
-    wc.M .= zero(T)
+# Ref： https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/welford_var_estimator.hpp
+mutable struct WelfordVar{T<:AbstractVecOrMat{<:AbstractFloat}} <: VarEstimator{T}
+    n     :: Int
+    n_min :: Int
+    μ     :: T
+    M     :: T
+    δ     :: T    # cache for diff
+    var   :: T    # cache for variance
 end
 
-function add_sample!(wc::WelfordCov, s::AbstractVector)
-    wc.n += 1
-    @unpack δ, μ, n, M = wc
-    δ .= s .- μ
-    μ .+= δ ./ n
-    M .+= (s .- μ) .* δ'
-end
+Base.show(io::IO, ::WelfordVar) = print(io, "WelfordVar")
 
-# Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/covar_adaptation.hpp
-function get_cov(wc::WelfordCov{T}) where {T<:AbstractFloat}
-    n, M = T(wc.n), wc.M
-    @assert n >= 2 "Cannot get variance with only one sample"
-    return (n / ((n + 5) * (n - 1))) .* M + T(1e-3) * (5 / (n + 5)) * LinearAlgebra.I
+function WelfordVar(
+    ::Type{T}, sz::Union{Tuple{Int},Tuple{Int,Int}}; 
+    n_min::Int=10, var=ones(T, sz)
+) where {T<:AbstractFloat}
+    return WelfordVar(0, n_min, zeros(T, sz), zeros(T, sz), zeros(T, sz), var)
```

Comment:
Again, without `n` and with different type parameters the natural form would be
```suggestion
function WelfordVar{T}(
	sz::Union{Tuple{Int},Tuple{Int,Int}}; 
    n_min::Int=10, var=ones(T, sz)
) where {T<:AbstractFloat}
    return WelfordVar(n_min, zeros(T, sz), zeros(T, sz), zeros(T, sz), var)
```